python/tvm/relay/quantize/quantize.py

```diff
@@ -373,3 +738,135 @@ def quantize(graph, params=None, dataset=None):
+        m.get_output(0, out_arr)
+        for o in range(0, len(output_collection)):
+            if output_collection[o] is None:
+                output_collection[o] = m.get_output(o+1).asnumpy()
+            else:
+                output_collection[o] = np.concatenate((output_collection[o], m.get_output(o+1).asnumpy()))
+        acc_top1.update(label, [mx.nd.array(out_arr.asnumpy())])
+        acc_top5.update(label, [mx.nd.array(out_arr.asnumpy())])
+        _, top1 = acc_top1.get()
+        _, top5 = acc_top5.get()
+
+        if not (i + 1) % log_iter:
+            nsamples = (i + 1) * batch_size
+            print('[{0:d} samples] evaluation: acc-top1={1:f} acc-top5={2:f}'.format(nsamples, top1, top5))
+
+        if (i+1)*batch_size >= early_stopping:
+            return top1, output_collection
+
+#def autoquantize(graph_callback, tr_data, tr_batch_fn, granularity='layer'):
+def autoquantize(graph, params, tr_data, tr_batch_fn, granularity='layer'):
+
+    import tvm
+    import copy
+    from tvm import relay
+    from tvm.relay import ir_pass
+
+    #graph, params = graph_callback()
+
+    graph = optimize(graph, params)
+    with qconfig(skip_k_conv=0,
+                 passthrough_bound=int(-1),
+                 nbit_input=8,
+                 nbit_weight=8,
+                 global_scale=8.0,
+                 dtype_input='int8',
+                 dtype_weight='int8',
+                 dtype_activation='int32',
+                 store_lowbit_output=True,
+                 debug_enabled_ops=None,
+                 granularity=granularity):
+        layout_map = tag_layout(graph)
+        graph = annotate(graph, layout_map)
+        annotated = copy.deepcopy(graph)
+        graph = ir_pass.infer_type(graph)
+        graph, profile_data = calibrate(graph, profile_mode=True, scales=None)
+        
+        free_vars = list(ir_pass.free_vars(graph))
+        graph = relay.Function(list(graph.params) + free_vars,
+                                graph.body, graph.ret_type,
+                                graph.type_params, graph.attrs)
+        additional_outputs = list()
+        metadata = list()
+        for hint, data, granularity, layout in profile_data:
+            additional_outputs.append(data)
+            metadata.append((hint, granularity, layout))
+        graph = relay.Function(graph.params,
+                                relay.expr.Tuple([graph.body]+additional_outputs))
+        target = 'llvm -mcpu=core-avx2'
+        #target = 'cuda'
+        with relay.build_config(opt_level=0):
+            graph, lib, params = relay.build(graph, target)
+            ctx = tvm.nd.context(target)
+
+        config = list()
+        print("calibrating...")
+        t1 = time.time()
+        top1, outputs = _evaluate(tr_data, tr_batch_fn, graph, lib, params, ctx, free_vars, early_stopping=64)
+        for i, output in enumerate(outputs):
+            config.append(_mse_chooser(output, granularity, metadata[i][-1]))
```

Comment:
`metadata` include the input meta, but `outputs`  ignore the input data. The output may mismatch its' metadata. Maybe we should replace `_mse_chooser(output, granularity, metadata[i][-1])` with `_mse_chooser(output, granularity, metadata[i+1][-1])`

Another question, Should we ignore the input data for quantization ?  `INPUT` type is included in quantization kind .

Reply:
We really only care that the length of profile_data matches the length of metadata
Type also does not specify parameters such as scale for calibration

Reply:
thanks for your reply. Maybe i haven't fully understand the quantization pass. that caused my doubt.